### PR TITLE
Implement S3 presigned_url API function

### PR DESCRIPTION
**Pull request by jordi-chacon**
Hi Ben,

this is a very early phase PR (i.e. I am not done yet). I'd very much like to get some early feedback from you before I proceed.

My initial idea was to change as little as possible from the existing code, simply add on. But then I found that `auth.ex` had a lot of bits and pieces of what I needed, so I was kind of forced to refactor it to avoid duplication. I hope you are OK with this.

I have not tested this, not even manually, I just got it to a state where it compiles and looks like it could work. I will probably continue tomorrow. But I just would like to get a thumbs up or thumbs down from you in regards to what I have done so far.

I also have a question for you: I see that function `ExAws.S3.Request.url` is used to construct S3 URLs. At first I thought I could use it, but then I saw this rule about the `bucket` containing a `.` which I didn't understand. Why this rule? For the `presigned_url` function, there is a `virtual_host` flag which indicates whether I should include the `bucket` in the hostname or in the path. So for now I chose to build the URL myself until I get a better understanding of the existing function and find out if I can reuse it or not.

Thanks a lot Ben!

### CODE REVIEW

**Comment by jordi-chacon** [comment]
Hi again Ben,
now I have completed the implementation and the testing of URL presigning for S3.

**Comment by benwilson512** [comment]
Hey thanks! I should be able to look through this over the weekend. I'll try to figure out why travis-ci with PRs doesn't work as well.

**Comment by benwilson512** [comment]
http://docs.travis-ci.com/user/pull-requests/ is why the tests won't pass, which makes sense. I should tag the tests that do actual requests to AWS and use the `TRAVIS_PULL_REQUEST` env variable to not run them.

**Comment by jordi-chacon** [comment]
Just wanted to mention that I have done two manual tests:
* Generate a presigned URL for `PUT`, and then upload a text file to S3 by curling such URL. (`curl -L --upload-file foo.txt "url"`)
* Generate a presigned URL for `GET`, and then fetch a text from S3 by curling such URL. (`curl -L "url"`)

File: lib/ex_aws/auth.ex [hunk 1]
```diff
@@ -97,26 +69,58 @@ defmodule ExAws.Auth do
     |> Keyword.keys
     |> Enum.join(";")
 
+    payload = case body do
+                nil -> "UNSIGNED-PAYLOAD"
+                _ -> ExAws.Auth.Utils.hash_sha256(body)
+              end
```
Comment (benwilson512) [inline lib/ex_aws/auth.ex]: As a minor coding style thing, I would prefer
```
payload = case body do
    nil -> "UNSIGNED-PAYLOAD"
    _ -> ExAws.Auth.Utils.hash_sha256(body)
end
```

This is the style used by the rest of the code base and I would prefer to keep that consistent.

File: lib/ex_aws/s3/client.ex [hunk 2]
```diff
@@ -411,6 +416,16 @@ defmodule ExAws.S3.Client do
...
 
   @doc """
+  Generates a pre-signed URL for this object.
+
+  Raises an ArgumentError if `:expires_in` exceeds one week.
+
+  When option param :virtual_host is `true`, the {#bucket} name will be used as
+  the hostname. This will cause the returned URL to be 'http' and not 'https'.
+  """
+  defcallback presigned_url(http_method :: atom, bucket :: binary, object :: binary, opts :: presigned_url_opts) :: binary
```
Comment (benwilson512) [inline lib/ex_aws/s3/client.ex]: Related to the comment on the implementation, I believe the return specs here should be `{:ok, binary} | {:error, binary}`

File: lib/ex_aws/s3/impl.ex [hunk 3]
```diff
@@ -441,8 +441,28 @@ defmodule ExAws.S3.Impl do
     |> Parsers.parse_list_parts
   end
 
+  def presigned_url(client, http_method, bucket, object, opts \\ []) do
+    expires_in = Keyword.get(opts, :expires_in, 3600)
+    virtual_host = Keyword.get(opts, :virtual_host, false)
+    raise_if_exceeds_one_week(expires_in)
```
Comment (benwilson512) [inline lib/ex_aws/s3/impl.ex]: Hey, so I would prefer if we did not raise here, and instead returned an {:error, reason} tuple. Correspondingly we should have a return of `{:ok, url}` on success.

**Comment by benwilson512** [comment]
Hey so other than the relatively minor tweaks this looks great. I really appreciate that you wrote comprehensive tests and typespec.

**Comment by jordi-chacon** [comment]
Thanks for the feedback Ben! I have now done the minor tweaks you requested.

**Comment by jordi-chacon** [comment]
I might have found a bug in my code regarding the `X-Amz-Security-Token`. Let's move this branch back to "not ready". I'll get back to you.

**Comment by jordi-chacon** [comment]
It is fixed now.

**Comment by benwilson512** [comment]
I'm calling this good to go, should publish it to hex later today. Thanks!

**Comment by jordi-chacon** [comment]
Excellent! Thank you!

**Pull request merged**